manual/src/refman/expr.etex

```diff
@@ -1292,58 +1268,85 @@ otherwise the default is nonvariant. This is also the case for
-  (b -. a) /. (c -. b);;
+let f x y =
+  let type t = {x: float; y: float} in
+  let norm p = p.x *. p.x +. p.y *. p.y in
+  norm {x; y}
+;;
 \end{caml_example}
 
-As a special case, @"assert false"@ is reduced to
-@'raise' '('@"Assert_failure ..."@')'@, which gives it a polymorphic
-type.  This means that it can be used in place of any expression (for
-example as a branch of any pattern-matching).  It also means that
-the @"assert false"@ ``assertions'' cannot be turned off by the
-"-noassert" option.
-%
-\index{Assertfailure\@\verb`Assert_failure`}
+The compiler checks that values created from this type do not escape this scope:
 
 \begin{caml_example}{toplevel}
-let min_known_nonempty = function
-  | [] -> assert false
-  | l -> List.hd (List.sort compare l);;
+let f x y =
+  let type t = {x: float; y: float} in
+  let twice p = {x = p.x +. p.x; y = p.y +. p.y} in
+  twice {x; y}
+[@@expect error];;
 \end{caml_example}
 
-\subsubsection*{sss:expr-lazy}{Lazy expressions}
-\ikwd{lazy\@\texttt{lazy}}
+Extensible types can also be introduced in this way. The example below exhibits
+a function that uses a local effect to convert an iterator into a sequence
+(``inversion of control'').
 
-The expression @"lazy" expr@ returns a value \var{v} of type "Lazy.t" that
-encapsulates the computation of @expr@.  The argument @expr@ is not
-evaluated at this point in the program.  Instead, its evaluation will
-be performed the first time the function "Lazy.force" is applied to the value
-\var{v}, returning the actual value of @expr@. Subsequent applications
-of "Lazy.force" to \var{v} do not evaluate @expr@ again. Applications
-of "Lazy.force" may be implicit through pattern matching (see~\ref{sss:pat-lazy}).
+\begin{caml_example}{toplevel}
+let invert (type a) ~(iter : (a -> unit) -> unit) : a Seq.t =
+  let type _ Effect.t += Yield : a -> unit Effect.t in
+  let yield v = Effect.perform (Yield v) in
+  fun () -> match iter yield with
+  | () -> Seq.Nil
+  | effect Yield v, k -> Seq.Cons (v, Effect.Deep.continue k)
+;;
+\end{caml_example}
+
+\subsubsection{sss:expr-let-exception}{Local exceptions}
+(Introduced in OCaml 4.04)
+
+It is possible to define local exceptions in expressions:
+@ "let" exception constr-decl "in" expr @ .
 
 \begin{caml_example}{toplevel}
-let lazy_greeter = lazy (print_string "Hello, World!\n");;
+let map_empty_on_negative f l =
+  let exception Negative in
+  let aux x = if x < 0 then raise Negative else f x in
+    try List.map aux l with Negative -> [];;
+\end{caml_example}
 
-Lazy.force lazy_greeter;;
+The syntactic scope of the exception constructor is the inner
+expression, but nothing prevents exception values created with this
+constructor from escaping this scope.  Two executions of the definition
+above result in two incompatible exception constructors (as for any
+exception definition). For instance:
+
+\begin{caml_example}{toplevel}
+let gen () = let exception A in A
+
+let () = assert(gen () = gen ());;
```

Comment:
Urh, it seems that github doesn't support commenting on unmodified parts of the code.

Anyway, the first sentence of the `let module` is now describing something less general than is supported (in fact, the base description is already incorrect, as it claims that `let module F(A : B) = ... in ...` is not valid, which it is).

This should presumably refer to some kind of `module_decl` grammar item.